dotnet/src/InternalUtilities/src/Type/TypeExtensions.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace System;
+
+/// <summary>
+/// Extensions methods for <see cref="System.Type"/>.
+/// </summary>
+internal static class TypeExtensions
+{
+    /// <summary>
+    /// Tries to get the result type from a generic parameter of type Task, Nullable, or ValueTask.
+    /// </summary>
+    /// <param name="returnType">Return type.</param>
+    /// <param name="resultType">The result type of the Nullable generic parameter.</param>
+    /// <returns><c>true</c> if the result type was successfully retrieved; otherwise, <c>false</c>.</returns>
+    public static bool TryGetGenericResultType(this Type? returnType, out Type resultType)
```

Comment:
@stephentoub is this the best way to do this?

Reply:
Chatted with Stephen offline -- Will leave as is for now since this doesn't impact the public API surface area - It's an implementation detail. 

He suggested either temporarily living with the overhead, or putting a cache in place, e.g. a static readonly ConcurrentDictionary that maps the input to the output, so that you only need to compute the answer for a given type once. Issue #4202 to track this